### PR TITLE
`<filesystem>`: `path` should be `range`

### LINKED ISSUE #2358: <filesystem>: `path`'s comparison operators are IF-NDR
**Describe the bug**

After #2000, `path`'s comparison operators are IF-NDR (ill-formed, no diagnostic required). See @Quuxplusone's [blogpost](https://quuxplusone.github.io/blog/2021/11/21/concepts-what-are-you-doing/) and @jwakely's [comment](https://www.reddit.com/r/cpp/comments/qz699h/is_path_convertible_to_string_view_a_war_story/hlmhzxr/) for a detailed analysis of similar issues in libc++ and libstdc++.

**Command-line test case**

Copied from Arthur O’Dwyer's blogpost.

```
D:\Temp>type repro.cpp
```
```cpp
#include <filesystem>
#include <ranges>

static_assert(std::ranges::range<std::filesystem::path>);
static_assert(std::ranges::range<const std::filesystem::path>);
```
```
D:\Temp>clang-cl /c /EHsc /W4 /std:c++latest /D_HAS_CXX23=1 repro.cpp
repro.cpp(5,1): error: static_assert failed
static_assert(std::ranges::range<const std::filesystem::path>);
^                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
repro.cpp(5,28): note: because 'const std::filesystem::path' does not satisfy 'range'
static_assert(std::ranges::range<const std::filesystem::path>);
                           ^
D:\repos\STL\out\\build\x64\out\inc\xutility(2233,9): note: because '::std::ranges::begin(__r)' would be invalid: no
      matching function for call to object of type 'const _Begin::_Cpo'
        _RANGES begin(__r);
        ^
D:\repos\STL\out\\build\x64\out\inc\yvals_core.h(1411,20): note: expanded from macro '_RANGES'
#define _RANGES    ::std::ranges::
                   ^
1 error generated.
```

**STL version**

https://github.com/microsoft/STL/commit/3c2fd04d441d46ec9d914d9cbb621a3bac96c3a5

### PR DESCRIPTION
**Pull request by fsb4000**
Fixes #2358

### CODE REVIEW

File: stl/inc/filesystem [hunk 1]
```diff
@@ -1434,6 +1432,16 @@ namespace filesystem {
         }
 
     private:
+        // operators ==, <==> and += have overloads with path and basic_string_view.
+        // clang-cl to attempt implicit conversion from const path& to wstring_view,
+        // and thus affect the result of range<const path>
+        // so we added the helpers to avoid the implicit conversios
+        inline int _Compare(const basic_string_view<value_type> _Other) const noexcept;
```
Comment (AlexGuteniev) [inline stl/inc/filesystem:1439]: ↕️ more space ↕️

Comment (AlexGuteniev) [reply stl/inc/filesystem:1439]: And `_NODISCARD`

File: stl/inc/filesystem [hunk 2]
```diff
@@ -1434,6 +1432,16 @@ namespace filesystem {
         }
 
     private:
+        // operators ==, <==> and += have overloads with path and basic_string_view.
+        // clang-cl to attempt implicit conversion from const path& to wstring_view,
+        // and thus affect the result of range<const path>
+        // so we added the helpers to avoid the implicit conversios
+        inline int _Compare(const basic_string_view<value_type> _Other) const noexcept;
+        _NODISCARD path& _PlusEqual(const path& _Added) { // concat _Added to native()
```
Comment (AlexGuteniev) [inline stl/inc/filesystem:1440]: `_PlusAssign`

Comment (AlexGuteniev) [reply stl/inc/filesystem:1440]: Or maybe `_Append`

**Comment by fsb4000** [comment]
I'm unsure: Should `_Compare` be defined after `path::iterator` is complete?
As far as I understood from the discussion this is necessary, but the tests passed without it...

File: tests/std/tests/P0896R4_ranges_test_machinery/test.compile.pass.cpp [hunk 3]
```diff
@@ -12,6 +13,10 @@ using namespace std;
 
 int main() {} // COMPILE-ONLY
 
+// GH-2358: <filesystem>: path's comparison operators are IF-NDR
```
Comment (CaseyCarter) [inline tests/std/tests/P0896R4_ranges_test_machinery/test.compile.pass.cpp:16]: This file is for testing the ranges test machinery defined in `<range_algorithm_support.hpp>`. I think this test would fit better somewhere in `P0896R4_ranges_range_machinery`.

File: stl/inc/filesystem [hunk 4]
```diff
@@ -1736,8 +1748,9 @@ namespace filesystem {
             const path& _Elem = *_B_iter;
 
             if (_Elem.empty()) { // skip empty element, N4810 29.11.7.5 [fs.path.itr]/4.4
-            } else if (_Elem == _Dot) { // skip filename elements that are dot, N4810 29.11.7.4.11 [fs.path.gen]/4.2
-            } else if (_Elem == _Dot_dot) {
+            } else if (_Elem == path(_Dot)) {
+                // skip filename elements that are dot, N4810 29.11.7.4.11 [fs.path.gen]/4.2
+            } else if (_Elem == path(_Dot_dot)) {
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/filesystem:1753]: OIC - we're avoiding the trigger condition despite that `_Path_iterator` is complete here so the trigger condition won't cause the bug, I think we're fine to revert this and avoid pessimizing codegen / debugging experience with the temporary `path`s.

File: stl/inc/filesystem [hunk 5]
```diff
@@ -1434,6 +1432,20 @@ namespace filesystem {
         }
 
     private:
+        // compare, operators ==, <==> and += have overloads with path and basic_string_view.
+        // clang-cl to attempt implicit conversion from const path& to wstring_view,
+        // and thus affect the result of range<const path>
+        // so we added the helpers to avoid the implicit conversions
+        _NODISCARD int _Compare(const basic_string_view<value_type> _Other) const noexcept {
+            return compare(_Other);
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/filesystem:1440]: I'll push a change for this since it's probably easier than explaining what I mean.

**Review by CaseyCarter — APPROVED** [review]
I pushed changes for my comments, which are a bit more invasive than I'd typically do in someone's branch without asking. Feel free to pushback if you disagree with anything.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for fixing `path` to behave properly after this surprising interaction with C++23! :heart_eyes_cat: :rocket: :white_check_mark:

**Pull request merged**